Model/ProductRepository.php

```diff
@@ -384,4 +386,25 @@ private function setExtensionAttributes($product, $storeId): void
 
         $product->setExtensionAttributes($extensionAttributes);
     }
-}
\ No newline at end of file
+
+    /**
+     * Function to remove the special_price and its related fields from the
+     * custom_attributes.
+     * If we don't remove them, these attributes may cause some errors when a
+     * special_price is set but it is equal to the retail_price.
+     *
+     * @param ProductInterface $product
+     * @return void
+     */
+    public function removeSpecialPriceCustomAttributes($product)
+    {
+        //Remove special_price attributes
+        $attributes = ['special_price', 'special_from_date', 'special_to_date'];
+        foreach ($attributes as $attribute) {
+            if (!isset($product[$attribute])) {
+                continue;
+            }
+            unset($product[$attribute]);
```

Comment:
In order to avoid the "continue" command, which is useful for some cases but, IMO, it should be avoided, we could just:

```suggestion
            if (isset($product[$attribute]))
                unset($product[$attribute]);
```